### DISCUSSION TITLE
Fragment caching in Rails

### DISCUSSION THREAD

**Discussion by HellRok** [yippee-fun/phlex#744] [Help]
I'm wondering if there's any way to use Rails fragment caching with Phlex?

I've got this basic view which tries to write to the cache but since the rendering happens later instead of being returned, it's caching `nil`

```ruby
class Home::IndexView < ApplicationView
  def initialize(posts:)
    @posts = posts
  end

  def view_template
    # Breadcrumbs
    content_for(:breadcrumbs) {
      plain "Home"
    }

    @posts.each do |post|
      helpers.cache post do
        render Posts::SummaryComponent.new(
          post: post
        )
      end
    end
  end
end
```

I had a brief look at the code for Phlex and didn't see anything obvious that matches what I'm trying to do. Have I missed something? Is there a different way to do fragment caching I'm not aware of?

**Comment by ccrisc** [comment]
In Rails to use fragment caching effectively with Phlex you should return the content as a string. So you need to capture the output of the Phlex component rendering and then cache it. You can try to use a buffer:



    @posts.each do |post|
      cached_post = helpers.cache(post) do
        capture do
          render Posts::SummaryComponent.new(post: post)
        end
      end

      text cached_post
    end

**Reply by HellRok** [reply]
Interestingly this isn't working for me. I'm trying with this code and just getting blank results:

```ruby
@posts.each do |post|
  cached_post = helpers.cache(post) do
    capture do
      render Posts::SummaryComponent.new(post: post)
    end
  end

  plain cached_post
end
```

The interesting part is that *this* does work (obviously without caching):

```ruby
@posts.each do |post|
  cached_post = capture do
    render Posts::SummaryComponent.new(post: post)
  end

  plain cached_post
end
```

**Comment by joeldrapper** [comment]
It’s not possible to use fragment caching with Phlex, because you would need to track all the dependencies. Even ActionView doesn’t do this very well. However, Phlex is so fast, it’s not necessary to use fragment caching. You can use caching to speed up database queries or API requests, while continuing to render views on the fly. The views won't be the bottleneck.
